File: stacks/python/base/code/database_models.tpl.py
"""
File: database_models.tpl.py
Purpose: SQLAlchemy model definitions with common patterns
Generated for: {{PROJECT_NAME}}
"""

from datetime import datetime
from typing import Optional, List
from sqlalchemy import Column, Integer, String, Boolean, DateTime, ForeignKey, Text, JSON
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship
from sqlalchemy.ext.asyncio import AsyncAttrs


class Base(AsyncAttrs, DeclarativeBase):
    """Base model with common functionality"""
    pass


class TimestampMixin:
    """Mixin for created/updated timestamps"""
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )


class SoftDeleteMixin:
    """Mixin for soft delete functionality"""
    deleted_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False)

    def soft_delete(self):
        self.deleted_at = datetime.utcnow()
        self.is_deleted = True


class User(Base, TimestampMixin, SoftDeleteMixin):
    """User model with authentication fields"""
    __tablename__ = "users"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    email: Mapped[str] = mapped_column(String(255), unique=True, nullable=False, index=True)
    username: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    password_hash: Mapped[str] = mapped_column(String(255), nullable=False)
    full_name: Mapped[Optional[str]] = mapped_column(String(100))
    avatar_url: Mapped[Optional[str]] = mapped_column(String(500))
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    is_verified: Mapped[bool] = mapped_column(Boolean, default=False)
    is_superuser: Mapped[bool] = mapped_column(Boolean, default=False)
    last_login_at: Mapped[Optional[datetime]] = mapped_column(DateTime)
    metadata_: Mapped[Optional[dict]] = mapped_column("metadata", JSON, default=dict)

    # Relationships
    sessions: Mapped[List["Session"]] = relationship(back_populates="user", lazy="selectin")
    posts: Mapped[List["Post"]] = relationship(back_populates="author", lazy="selectin")

    def __repr__(self):
        return f"<User(id={self.id}, email={self.email})>"


class Session(Base, TimestampMixin):
    """User session for token management"""
    __tablename__ = "sessions"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"))
    token_hash: Mapped[str] = mapped_column(String(255), unique=True, nullable=False, index=True)
    expires_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    ip_address: Mapped[Optional[str]] = mapped_column(String(45))
    user_agent: Mapped[Optional[str]] = mapped_column(String(500))
    is_revoked: Mapped[bool] = mapped_column(Boolean, default=False)

    # Relationships
    user: Mapped["User"] = relationship(back_populates="sessions")


class Post(Base, TimestampMixin, SoftDeleteMixin):
    """Example content model"""
    __tablename__ = "posts"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    author_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"))
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    slug: Mapped[str] = mapped_column(String(220), unique=True, nullable=False, index=True)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    excerpt: Mapped[Optional[str]] = mapped_column(String(500))
    status: Mapped[str] = mapped_column(String(20), default="draft")  # draft, published, archived
    published_at: Mapped[Optional[datetime]] = mapped_column(DateTime)
    view_count: Mapped[int] = mapped_column(Integer, default=0)
    metadata_: Mapped[Optional[dict]] = mapped_column("metadata", JSON, default=dict)

    # Relationships
    author: Mapped["User"] = relationship(back_populates="posts")
    tags: Mapped[List["Tag"]] = relationship(
        secondary="post_tags", back_populates="posts", lazy="selectin"
    )


class Tag(Base, TimestampMixin):
    """Tag model for categorization"""
    __tablename__ = "tags"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    slug: Mapped[str] = mapped_column(String(60), unique=True, nullable=False, index=True)
    description: Mapped[Optional[str]] = mapped_column(String(200))
    color: Mapped[Optional[str]] = mapped_column(String(7))  # Hex color

    # Relationships
    posts: Mapped[List["Post"]] = relationship(
        secondary="post_tags", back_populates="tags", lazy="selectin"
    )


# Association table for many-to-many
from sqlalchemy import Table
post_tags = Table(
    "post_tags",
    Base.metadata,
    Column("post_id", Integer, ForeignKey("posts.id", ondelete="CASCADE"), primary_key=True),
    Column("tag_id", Integer, ForeignKey("tags.id", ondelete="CASCADE"), primary_key=True),
)


class AuditLog(Base):
    """Audit log for tracking changes"""
    __tablename__ = "audit_logs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    action: Mapped[str] = mapped_column(String(50), nullable=False)  # create, update, delete
    entity_type: Mapped[str] = mapped_column(String(50), nullable=False)
    entity_id: Mapped[str] = mapped_column(String(50), nullable=False)
    old_values: Mapped[Optional[dict]] = mapped_column(JSON)
    new_values: Mapped[Optional[dict]] = mapped_column(JSON)
    ip_address: Mapped[Optional[str]] = mapped_column(String(45))
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

File: stacks/go/base/code/database_models.tpl.go
// File: database_models.tpl.go
// Purpose: GORM model definitions with common patterns
// Generated for: {{PROJECT_NAME}}

package models

import (
	"time"

	"gorm.io/gorm"
)

// Base model with common fields
type BaseModel struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
}

// User model with authentication fields
type User struct {
	BaseModel
	Email        string     `gorm:"size:255;uniqueIndex;not null" json:"email"`
	Username     string     `gorm:"size:50;uniqueIndex;not null" json:"username"`
	PasswordHash string     `gorm:"size:255;not null" json:"-"`
	FullName     string     `gorm:"size:100" json:"full_name,omitempty"`
	AvatarURL    string     `gorm:"size:500" json:"avatar_url,omitempty"`
	IsActive     bool       `gorm:"default:true" json:"is_active"`
	IsVerified   bool       `gorm:"default:false" json:"is_verified"`
	IsSuperuser  bool       `gorm:"default:false" json:"is_superuser"`
	LastLoginAt  *time.Time `json:"last_login_at,omitempty"`
	Metadata     JSON       `gorm:"type:jsonb" json:"metadata,omitempty"`

	// Relationships
	Sessions []Session `gorm:"foreignKey:UserID" json:"sessions,omitempty"`
	Posts    []Post    `gorm:"foreignKey:AuthorID" json:"posts,omitempty"`
}

// TableName specifies the table name for User
func (User) TableName() string {
	return "users"
}

// Session model for token management
type Session struct {
	BaseModel
	UserID    uint      `gorm:"index;not null" json:"user_id"`
	TokenHash string    `gorm:"size:255;uniqueIndex;not null" json:"-"`
	ExpiresAt time.Time `gorm:"not null" json:"expires_at"`
	IPAddress string    `gorm:"size:45" json:"ip_address,omitempty"`
	UserAgent string    `gorm:"size:500" json:"user_agent,omitempty"`
	IsRevoked bool      `gorm:"default:false" json:"is_revoked"`

	// Relationships
	User User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

// TableName specifies the table name for Session
func (Session) TableName() string {
	return "sessions"
}

// Post model for content
type Post struct {
	BaseModel
	AuthorID    uint       `gorm:"index;not null" json:"author_id"`
	Title       string     `gorm:"size:200;not null" json:"title"`
	Slug        string     `gorm:"size:220;uniqueIndex;not null" json:"slug"`
	Content     string     `gorm:"type:text;not null" json:"content"`
	Excerpt     string     `gorm:"size:500" json:"excerpt,omitempty"`
	Status      string     `gorm:"size:20;default:'draft'" json:"status"` // draft, published, archived
	PublishedAt *time.Time `json:"published_at,omitempty"`
	ViewCount   int        `gorm:"default:0" json:"view_count"`
	Metadata    JSON       `gorm:"type:jsonb" json:"metadata,omitempty"`

	// Relationships
	Author User  `gorm:"foreignKey:AuthorID" json:"author,omitempty"`
	Tags   []Tag `gorm:"many2many:post_tags" json:"tags,omitempty"`
}

// TableName specifies the table name for Post
func (Post) TableName() string {
	return "posts"
}

// Tag model for categorization
type Tag struct {
	BaseModel
	Name        string `gorm:"size:50;uniqueIndex;not null" json:"name"`
	Slug        string `gorm:"size:60;uniqueIndex;not null" json:"slug"`
	Description string `gorm:"size:200" json:"description,omitempty"`
	Color       string `gorm:"size:7" json:"color,omitempty"` // Hex color

	// Relationships
	Posts []Post `gorm:"many2many:post_tags" json:"posts,omitempty"`
}

// TableName specifies the table name for Tag
func (Tag) TableName() string {
	return "tags"
}

// AuditLog for tracking changes
type AuditLog struct {
	ID         uint      `gorm:"primarykey" json:"id"`
	UserID     *uint     `gorm:"index" json:"user_id,omitempty"`
	Action     string    `gorm:"size:50;not null" json:"action"` // create, update, delete
	EntityType string    `gorm:"size:50;not null" json:"entity_type"`
	EntityID   string    `gorm:"size:50;not null" json:"entity_id"`
	OldValues  JSON      `gorm:"type:jsonb" json:"old_values,omitempty"`
	NewValues  JSON      `gorm:"type:jsonb" json:"new_values,omitempty"`
	IPAddress  string    `gorm:"size:45" json:"ip_address,omitempty"`
	CreatedAt  time.Time `json:"created_at"`

	// Relationships
	User *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

// TableName specifies the table name for AuditLog
func (AuditLog) TableName() string {
	return "audit_logs"
}

// JSON type for JSONB fields
type JSON map[string]interface{}

// AutoMigrate runs migrations for all models
func AutoMigrate(db *gorm.DB) error {
	return db.AutoMigrate(
		&User{},
		&Session{},
		&Post{},
		&Tag{},
		&AuditLog{},
	)
}

// Hooks

// BeforeCreate hook for User
func (u *User) BeforeCreate(tx *gorm.DB) error {
	if u.Metadata == nil {
		u.Metadata = make(JSON)
	}
	return nil
}

// BeforeUpdate hook for Post - update slug if title changed
func (p *Post) BeforeCreate(tx *gorm.DB) error {
	if p.Slug == "" && p.Title != "" {
		p.Slug = generateSlug(p.Title)
	}
	return nil
}

// Helper function to generate slug (implement as needed)
func generateSlug(title string) string {
	// Implement slug generation logic
	return title // Placeholder
}
